### PR TITLE
Added RandomWidth as native KerasCV preprocessing layer

### LINKED ISSUE #469: Migrate RandomWidth KPL to Keras_CV

### PR DESCRIPTION
**Pull request by suvadityamuk**
# What does this PR do?

Fixes #469 and adds RandomWidth layer to be used for preprocessing. Necessary tests for the code have also been added with all of them passing. Ported/Migrated from `keras/layers/preprocessing/image_preprocessing.py` and its respective tests.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/keras-team/keras-cv/blob/master/.github/CONTRIBUTING.md),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [Link to discussion issue for RandomWidth](https://github.com/keras-team/keras-cv/issues/469)
- [Link to discussion issue for RandomHeight](https://github.com/keras-team/keras-cv/issues/468)
- [X] Did you write any new necessary tests?
- [ ] If this adds a new model, can you run a few training steps on TPU in Colab to ensure that no XLA incompatible OP are used?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@LukeWood @tanzhenyu

### CODE REVIEW

File: keras_cv/layers/preprocessing/random_height.py
Comment (tanzhenyu) [inline keras_cv/layers/preprocessing/random_height.py:14]: tf.image.ResizeMethod is already string based, prefer using those

File: keras_cv/layers/preprocessing/random_height.py
Comment (tanzhenyu) [inline keras_cv/layers/preprocessing/random_height.py:21]: I *think* the canonical description would start with a one-liner describing what the layer does, and then blank, and then another paragraph on what is the behavior of this layer

File: keras_cv/layers/preprocessing/random_height.py
Comment (tanzhenyu) [inline keras_cv/layers/preprocessing/random_height.py:23]: maybe add detail as to the behavior for intra-batch? Is is all samples within a batch using the same height / width, or randomness is applied intra-batch?

**Review by tanzhenyu — COMMENTED** [review]
Thanks for the PR!
Also I wonder who'd have the 1000th PR :-)

File: keras_cv/layers/preprocessing/random_height.py
Comment (suvadityamuk) [reply keras_cv/layers/preprocessing/random_height.py:14]: Will do!

**Comment by suvadityamuk** [comment]
> Thanks for the PR! Also I wonder who'd have the 1000th PR :-)

I wish I could have broken these PRs and made the 1000th!

File: keras_cv/layers/preprocessing/random_height.py
Comment (bhack) [reply keras_cv/layers/preprocessing/random_height.py:23]: Does this imply to use ragged tensor if not padded?

File: keras_cv/layers/preprocessing/random_height.py
Comment (suvadityamuk) [reply keras_cv/layers/preprocessing/random_height.py:21]: Actually, this is a direct copy of the KPL docstring. I'll customize it out specific to KerasCV

File: keras_cv/layers/preprocessing/random_height.py
Comment (suvadityamuk) [reply keras_cv/layers/preprocessing/random_height.py:23]: Alright, shall do

**Comment by suvadityamuk** [comment]
> > > Thanks for the PR! Also I wonder who'd have the 1000th PR :-)
> > 
> > 
> > I wish I could have broken these PRs and made the 1000th!
> 
> You can still make it :-)

I was trying to, but I do not know how I could do so. Any pointers :sweat_smile: ?

File: keras_cv/layers/preprocessing/random_height.py
Comment (suvadityamuk) [reply keras_cv/layers/preprocessing/random_height.py:23]: > Does this imply to use ragged tensor if not padded?

I think it might, because how else would the different tensor configurations be handled?

File: keras_cv/layers/preprocessing/random_height.py
Comment (LukeWood) [reply keras_cv/layers/preprocessing/random_height.py:14]: Lets use the utility from my PR for Resizing.  `keras_cv.utils.get_interpolation()`.

Comment (suvadityamuk) [reply keras_cv/layers/preprocessing/random_height.py:14]: Alright, will do that instead!

**Comment by suvadityamuk** [comment]
> > > > Thanks for the PR! Also I wonder who'd have the 1000th PR :-)
> > > 
> > > 
> > > I wish I could have broken these PRs and made the 1000th!
> > 
> > 
> > You can still make it :-)
> 
> I was trying to, but I do not know how I could do so. Any pointers sweat_smile ?

Found a way to using some very smart cherry-picking, thank god for StackOverflow and GitHub after all

**Comment by suvadityamuk** [comment]
Oh well, seems I'll make the 1001th PR instead, got beat to the punch haha!

**Comment by suvadityamuk** [comment]
Please check the 2nd PR here 

https://github.com/keras-team/keras-cv/pull/1001

File: keras_cv/models/generative/stable_diffusion/stable_diffusion.py
Comment (LukeWood) [inline keras_cv/models/generative/stable_diffusion/stable_diffusion.py:426]: looks like a bad rebase in stable diffusion.

Comment (suvadityamuk) [reply keras_cv/models/generative/stable_diffusion/stable_diffusion.py:426]: Ohh, my word. Let me check that right away and get back to you on this.

Comment (suvadityamuk) [reply keras_cv/models/generative/stable_diffusion/stable_diffusion.py]: Please check now

**Comment by suvadityamuk** [comment]
Hi, folks! Any updates on the review? @LukeWood @tanzhenyu @bhack

File: examples/layers/preprocessing/classification/random_width_demo.py
Comment (tanzhenyu) [inline examples/layers/preprocessing/classification/random_width_demo.py:4]: no need for the demo

File: keras_cv/layers/preprocessing/random_width.py
Comment (tanzhenyu) [inline keras_cv/layers/preprocessing/random_width.py:11]: I'm still a little confused on why not reusing tf.image.RESIZE_METHOD

File: keras_cv/layers/preprocessing/random_width_test.py
Comment (tanzhenyu) [inline keras_cv/layers/preprocessing/random_width_test.py:53]: why need to disable formating?

**Review by tanzhenyu — CHANGES_REQUESTED** [review]
Overall LGTM, some minor comments

**Comment by suvadityamuk** [comment]
@LukeWood @bhack @tanzhenyu Hi, any updates on the review?

File: keras_cv/layers/preprocessing/random_width.py
Comment (LukeWood) [inline keras_cv/layers/preprocessing/random_width.py:108]: I don't think this actually works because RandomWidth() operates on a batch-wise basis.  This looks like it will operate on a sample-wise basis.

Perhaps we can break API compat and return on a sample-wise basis, but we'd need to ask @fchollet.

Comment (LukeWood) [reply keras_cv/layers/preprocessing/random_width.py:108]: ah wait, nevermind you're only calling it once in batch_augment.  apologies

Comment (suvadityamuk) [reply keras_cv/layers/preprocessing/random_width.py:110]: So, is this fine as is?

**Comment by suvadityamuk** [comment]
You may have to run the workflow again since I just pushed a new commit that fixes the linting errors. Sorry, just noticed :sweat_smile:
Thanks!

File: keras_cv/layers/preprocessing/random_width_test.py
Comment (LukeWood) [reply keras_cv/layers/preprocessing/random_width_test.py:53]: +1, any reason to disable this?

File: keras_cv/layers/preprocessing/random_width_test.py
Comment (LukeWood) [inline keras_cv/layers/preprocessing/random_width_test.py:115]: you can delete this test, we don't usually test non-user-facing APIs

**Review by LukeWood — APPROVED** [review]
looks good, some minor changes

File: keras_cv/layers/preprocessing/random_width_test.py
Comment (suvadityamuk) [reply keras_cv/layers/preprocessing/random_width_test.py:115]: Will do

File: keras_cv/layers/preprocessing/random_width_test.py
Comment (suvadityamuk) [reply keras_cv/layers/preprocessing/random_width_test.py:53]: No, actually. Will remove it

**Comment by suvadityamuk** [comment]
Please check now, thanks!

**Comment by divyashreepathihalli** [comment]
@suvadityamuk  - Thank you for the PR!
Can you please add a demo file for RandomWidth?
here are some example
- classification demo - https://github.com/keras-team/keras-cv/blob/master/examples/layers/preprocessing/classification/random_crop_demo.py
- Bounding box demo - https://github.com/keras-team/keras-cv/blob/master/examples/layers/preprocessing/bounding_box/random_rotation_demo.py

Also please add the demo output here.

**Review by divyashreepathihalli — CHANGES_REQUESTED** [review]
@suvadityamuk - Thank you for the PR!
Can you please add a demo file for RandomWidth?
here are some examples

classification demo - https://github.com/keras-team/keras-cv/blob/master/examples/layers/preprocessing/classification/random_crop_demo.py
Bounding box demo - https://github.com/keras-team/keras-cv/blob/master/examples/layers/preprocessing/bounding_box/random_rotation_demo.py
Also please add the demo output here.

File: keras_cv/layers/preprocessing/random_height.py
Comment (LukeWood) [reply keras_cv/layers/preprocessing/random_height.py]: @suvadityamuk is this now resolved?  Make sure to click "resolve conversation" after fixing issues!

**Review by LukeWood — APPROVED** [review]
This looks good to me, but lets wait for an additional approval from @tanzhenyu as he had some comments as well.

**Comment by LukeWood** [comment]
Closing due to inactivity.  Thanks for the PR~

**Pull request closed without merging**
